src/optimalcontrol/include/iDynTree/Integrators/FixedStepIntegrator.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2014,2018 Fondazione Istituto Italiano di Tecnologia
+ * Authors: Francesco Romano, Stefano Dafarra
+ * CopyPolicy: Released under the terms of the LGPLv2.1 or later, see LGPL.TXT
+ *
+ * Originally developed for Prioritized Optimal Control (2014)
+ * Refactored in 2018.
+ * Design inspired by
+ * - ACADO toolbox (http://acado.github.io)
+ * - ADRL Control Toolbox (https://adrlab.bitbucket.io/ct/ct_doc/doc/html/index.html)
+ */
+
+#ifndef IDYNTREE_OPTIMALCONTROL_FIXEDSTEPINTEGRATOR_H
+#define IDYNTREE_OPTIMALCONTROL_FIXEDSTEPINTEGRATOR_H
+
+#include <iDynTree/Integrator.h>
+#include <iDynTree/Core/VectorDynSize.h>
+
+namespace iDynTree {
+    namespace optimalcontrol {
+
+        class DynamicalSystem;
+
+        namespace integrators {
+
+        /**
+         * @warning This class is still in active development, and so API interface can change between iDynTree versions.
+         * \ingroup iDynTreeExperimental
+         */
+
+            class FixedStepIntegrator : public Integrator
```

Comment:
Can you suggest me how to properly benchmark this?

Reply:
I think @traversaro did some tests some time ago on Vectors and similar.

A test can be to try to call the method in two different situations:
- with a pointer (this will actually call the virtual function)
- calling directly the object method (i.e. not a pointer. The object itself)

Of course, put the calls in some big for loops to have an average.

Reply:
Does it depends on the method called? Actually when solving an ``OptimalControlProblem`` I don't call the ``integrate`` method but rather the ``evaluateCollocationConstraint`` method which is usually implemented in the most derived class. Does this change something?

Reply:
Btw, I did this test:
```cpp
    std::shared_ptr<Integrator> ptr;
    std::shared_ptr<ForwardEuler> direct = std::make_shared<ForwardEuler>(dynamicalSystem);
    ptr = direct;

    clock_t initT, endT;
    initT = clock();
    for (int i = 0; i < 10000; ++i) {
        ASSERT_IS_TRUE(ptr->evaluateCollocationConstraint(0.0, c1, c2, dT, v3));
    }
    endT = clock();
    std::cerr << "Elapsed time (ptr): " <<  static_cast<double>(endT - initT) / CLOCKS_PER_SEC * 1000.0 <<" ms."<<std::endl;

    initT = clock();
    for (int i = 0; i < 10000; ++i) {
        ASSERT_IS_TRUE((*direct).evaluateCollocationConstraint(0.0, c1, c2, dT, v3));
    }
    endT = clock();
    std::cerr << "Elapsed time (direct): " <<  static_cast<double>(endT - initT) / CLOCKS_PER_SEC * 1000.0 <<" ms."<<std::endl;
```
Compiled in ``DEBUG`` I got this result after three experiments (in a row):
```
Starting /home/sdafarra/Software/robotology-superbuild/build/robotology/iDynTree/bin/IntegratorsUnitTest...
Elapsed time (ptr): 12.482 ms.
Elapsed time (direct): 13.573 ms.
/home/sdafarra/Software/robotology-superbuild/build/robotology/iDynTree/bin/IntegratorsUnitTest exited with code 0

Starting /home/sdafarra/Software/robotology-superbuild/build/robotology/iDynTree/bin/IntegratorsUnitTest...
Elapsed time (ptr): 12.435 ms.
Elapsed time (direct): 12.692 ms.
/home/sdafarra/Software/robotology-superbuild/build/robotology/iDynTree/bin/IntegratorsUnitTest exited with code 0

Starting /home/sdafarra/Software/robotology-superbuild/build/robotology/iDynTree/bin/IntegratorsUnitTest...
Elapsed time (ptr): 12.145 ms.
Elapsed time (direct): 12.854 ms.
/home/sdafarra/Software/robotology-superbuild/build/robotology/iDynTree/bin/IntegratorsUnitTest exited with code 0
```

Reply:
I would do something different.
First of all, compile in Release. Performance benchmark should be computed on the optimised code, not the original one.
Second, do not use a pointer for the derived-class test, i.e.
```cpp
// pseudocode.. I do not remember the classes
std::shared_ptr<Integrator> baseClass = std::make_shared<ForwardEuler>(dynamicalSystem);

// Do your tests here on baseClass, i.e. baseClass->callTheMethod()

ForwardEuler euler(dynamicalSystem)
// Do your tests here on euler, i.e. euler.callTheMethod()

```

Can you try with the above suggestions?

Reply:
Yep sure. I compiled in Release, using a direct instantiation as you suggested. The results are still unclear. I thought that also they way the for loops are handled may lead to different results (which may explain why the first for loop is faster). Thus, I modified the test into this:
```cpp
    std::shared_ptr<Integrator> ptr = std::make_shared<ForwardEuler>(dynamicalSystem);
    ForwardEuler direct(dynamicalSystem);

    clock_t initT, endT, sumDirect = 0, sumPtr = 0;

    ASSERT_IS_TRUE(direct.evaluateCollocationConstraint(0.0, c1, c2, dT, v3)); //for eventual memory allocation
    ASSERT_IS_TRUE(ptr->evaluateCollocationConstraint(0.0, c1, c2, dT, v3));

    for (int i = 0; i < 10000; ++i) {
        initT = clock();
        ASSERT_IS_TRUE(direct.evaluateCollocationConstraint(0.0, c1, c2, dT, v3));
        endT = clock();
        sumDirect += (endT-initT);

        initT = clock();
        ASSERT_IS_TRUE(ptr->evaluateCollocationConstraint(0.0, c1, c2, dT, v3));
        endT = clock();
        sumPtr += (endT-initT);
    }
    std::cerr << "Elapsed time (direct): " <<  static_cast<double>(sumDirect) *1000 / CLOCKS_PER_SEC <<" ms."<<std::endl;
    std::cerr << "Elapsed time (ptr): " <<  static_cast<double>(sumPtr) *1000 / CLOCKS_PER_SEC <<" ms."<<std::endl; //Multiplying first to avoid decimation issues
```

Running the tests in a terminal (to move ``QtCreator`` out of the equation), I got this:
```
sdafarra@iiticublap104:~/Software/robotology-superbuild/build/robotology/iDynTree/bin (master)$ ./IntegratorsUnitTest 
Elapsed time (direct): 15.006 ms.
Elapsed time (ptr): 15.013 ms.
sdafarra@iiticublap104:~/Software/robotology-superbuild/build/robotology/iDynTree/bin (master)$ ./IntegratorsUnitTest 
Elapsed time (direct): 19.454 ms.
Elapsed time (ptr): 19.456 ms.
sdafarra@iiticublap104:~/Software/robotology-superbuild/build/robotology/iDynTree/bin (master)$ ./IntegratorsUnitTest 
Elapsed time (direct): 19.12 ms.
Elapsed time (ptr): 18.99 ms.
sdafarra@iiticublap104:~/Software/robotology-superbuild/build/robotology/iDynTree/bin (master)$ ./IntegratorsUnitTest 
Elapsed time (direct): 18.548 ms.
Elapsed time (ptr): 18.46 ms.
sdafarra@iiticublap104:~/Software/robotology-superbuild/build/robotology/iDynTree/bin (master)$ ./IntegratorsUnitTest 
Elapsed time (direct): 4.457 ms.
Elapsed time (ptr): 4.397 ms.
```

Reply:
`ASSERT_XX` is not typedeffed in release mode, isn't it?

BTW: I preferred the two separate for with a single clock measure

Reply:
I restored the for loop as before and removed the check on the output value:
```cpp
    ASSERT_IS_TRUE(direct.evaluateCollocationConstraint(0.0, c1, c2, dT, v3)); //for eventual memory allocation
    ASSERT_IS_TRUE(ptr->evaluateCollocationConstraint(0.0, c1, c2, dT, v3));

    initT = clock();
    for (int i = 0; i < 10000; ++i) {
        ptr->evaluateCollocationConstraint(0.0, c1, c2, dT, v3);
    }
    endT = clock();
    std::cerr << "Elapsed time (ptr): " <<  static_cast<double>(endT - initT) / CLOCKS_PER_SEC * 1000.0 <<" ms."<<std::endl;

    initT = clock();
    for (int i = 0; i < 10000; ++i) {
        direct.evaluateCollocationConstraint(0.0, c1, c2, dT, v3);
    }
    endT = clock();
    std::cerr << "Elapsed time (direct): " <<  static_cast<double>(endT - initT) / CLOCKS_PER_SEC * 1000.0 <<" ms."<<std::endl;
```

Here the results:
```
sdafarra@iiticublap104:~/Software/robotology-superbuild/build/robotology/iDynTree/bin (master)$ ./IntegratorsUnitTest 
Elapsed time (ptr): 5.582 ms.
Elapsed time (direct): 5.607 ms.
sdafarra@iiticublap104:~/Software/robotology-superbuild/build/robotology/iDynTree/bin (master)$ ./IntegratorsUnitTest 
Elapsed time (ptr): 5.568 ms.
Elapsed time (direct): 5.59 ms.
sdafarra@iiticublap104:~/Software/robotology-superbuild/build/robotology/iDynTree/bin (master)$ ./IntegratorsUnitTest 
Elapsed time (ptr): 0.436 ms.
Elapsed time (direct): 0.462 ms.
sdafarra@iiticublap104:~/Software/robotology-superbuild/build/robotology/iDynTree/bin (master)$ ./IntegratorsUnitTest 
Elapsed time (ptr): 5.581 ms.
Elapsed time (direct): 5.396 ms.
sdafarra@iiticublap104:~/Software/robotology-superbuild/build/robotology/iDynTree/bin (master)$ ./IntegratorsUnitTest 
Elapsed time (ptr): 4.769 ms.
Elapsed time (direct): 5.591 ms.
sdafarra@iiticublap104:~/Software/robotology-superbuild/build/robotology/iDynTree/bin (master)$ ./IntegratorsUnitTest 
Elapsed time (ptr): 0.433 ms.
Elapsed time (direct): 0.55 ms.
```
Inverting the two loops gives similar results.
I guess the variability may depend on the assert and the calls to ``clock()``.

Reply:
Really interesting. 
I think that for now we can be happy. We can think of some better tests later. I don't know if Dtrace  has some tools to test this particular case.